### ISSUE TITLE
[layout-dit] Convert CGB-DM to a diffusers pipeline

### ISSUE THREAD

**Issue by shunk031** [creative-graphic-design/design-generators#15]
Convert CGB-DM / LayoutDiT (arXiv 2024) to a diffusers pipeline so that the diffusion-transformer layout generator can be loaded with `from_pretrained` once the authors' checkpoint files are obtained or retrained.

## Approach

- Implement with diffusers: `CgbDmTransformer(ModelMixin, ConfigMixin)` for `vendor/layout-dit/cgbdm/layout_model.py::LayoutModel`, `CgbDmScheduler(SchedulerMixin, ConfigMixin)` for the cosine DDPM plus DDIM/refinement schedules in `vendor/layout-dit/cgbdm/diffusion.py::Diffusion`, and `CgbDmPipeline(DiffusionPipeline)` for image/saliency preprocessing and layout rendering.
- Preserve the key config values from `configs/{pku,cgl}.yaml`: `num_class` 4 for PKU and 5 for CGL, `max_elem=16`, `d_model=512`, `n_head=8`, `n_layers=4`, `feature_dim=1024`, `num_timesteps=1000`, `ddim_num_steps=100`, and task modes `uncond`, `c`, `cwh`, `complete`, and refinement paths.
- Use `scripts/test.py` and `scripts/run_single_image.py` as reference inference entry points; they load `torch.load(opt.check_path)` as a raw `LayoutModel` state dict, call `diffusion_model.model.load_state_dict(weights)`, then generate via `reverse_ddim`, `conditional_reverse_ddim`, or `refinement_reverse_ddim`.
- Weight distribution needs correction against issue #2: the local README points `Download pre-trained weights` to `https://github.com/yuli0103/CGB-DM`, that URL redirects/aliases to the same LayoutDiT repository, GitHub Releases are empty, and the project page has no additional weight/checkpoint link; only the dataset Drive URL `https://drive.google.com/file/d/1aT_QuLhmZdhLKGZbZgtKfrNfqeHKPqIo/view?usp=drive_link` is explicit.
- Keep dependencies isolated but relatively modern: README lists Python 3.9 and `requirements.txt` pins `torch==2.3.1`, `torchvision==0.18.1`, `transformers==4.42.4`, `huggingface-hub==0.24.0`, `opencv-python==4.10.0.84`, `einops`, `natsort`, and saliency/inpainting code also uses ModelScope LaMa plus ISNet/BASNet weights.
- Determinism is supported by `transformers.set_seed(seed)` in training/test/single-image scripts, but DDIM/noise calls still need a pipeline-level `torch.Generator` and parity fixtures for the rendered class/bbox outputs.
- License is Apache-2.0 in `vendor/layout-dit/LICENSE`.
- The issue #2 API should stay diffusers: the model is a continuous layout diffusion transformer with explicit scheduler state and sampling loops, not a one-shot autoregressive/GAN model.

## Steps

- [ ] Create `models/cgb-dm/` with diffusers dependencies and a `vendor` extra for saliency/inpainting support.
- [ ] Implement `CgbDmTransformer`, `CgbDmScheduler`, `CgbDmImageProcessor`, and `CgbDmPipeline` with PKU/CGL config presets.
- [ ] Before conversion, resolve the missing pretrained checkpoint URL by contacting upstream or retraining; if no public weights are available, open a blocker note and do not claim ready-heavy weight availability.
- [ ] Generate reference outputs with `scripts/test.py` or `scripts/run_single_image.py` once a checkpoint is available, using fixed `--seed` and saved saliency boxes.
- [ ] Write a conversion script for raw `LayoutModel` state dicts and saliency detector metadata, and save scheduler buffers from config instead of checkpoint.
- [ ] Add no-weight scheduler/model-shape tests and parity tests comparing DDIM sample trajectories for a fixed `torch.Generator`.

## References

- Paper: https://arxiv.org/abs/2407.15233
- Weights: not found in README/project page/GitHub Releases; dataset link is https://drive.google.com/file/d/1aT_QuLhmZdhLKGZbZgtKfrNfqeHKPqIo/view?usp=drive_link
- Original implementation: vendor/layout-dit (Apache-2.0)

**Comment by shunk031** [comment]
Weights availability check result: no public CGB-DM generator checkpoints exist. This issue is being moved from milestone `v0.3: Ready (heavy)` to `v0.5: Train ourselves`; it can move back once upstream publishes checkpoints.

- Upstream repository: https://github.com/yuli0103/LayoutDiT (https://github.com/yuli0103/CGB-DM resolves to the same repo). Code license Apache-2.0, no releases, and the repo tree contains no `.pt`/`.pth` model weights — only small output metadata files (~16–113 KB).
- The README's "Download pre-trained weights" link points back to the repository itself, not to checkpoint files.
- Hugging Face `lyabc/CGB-DM` is public but empty: siblings are only `.gitattributes` and `README.md`; `pku.pt`/`cgl.pt` resolve 404.
- Upstream issue #1: the author said on 2024-07-25 "We plan to release them soon" and on 2024-09-26 that enhanced pretrained weights would follow after acceptance; issues #6 and #10 also ask for the weights. No maintainer reply with a URL to date.
- Only the dataset is public: Google Drive file `Dataset.zip` (~13 GB; organized PKU/CGL data, inpainted images, saliency maps, labels), which supports a train-ourselves path.

<details>
<summary>Sources checked</summary>

- design-generators issue #15 and the issue #2 target table (`Weights = Unconfirmed (no public URL found)`).
- GitHub API for `yuli0103/LayoutDiT`: default branch `master`, license Apache-2.0, releases `[]`; recursive tree scanned for checkpoint extensions.
- README links: data Google Drive id `1aT_QuLhmZdhLKGZbZgtKfrNfqeHKPqIo` (public, `Dataset.zip`, virus-scan page reports 13G); weights link points to the repo.
- Hugging Face Hub API: `lyabc/CGB-DM` siblings list; searches for `LayoutDiT` / `CGB-DM` found no other repos.
- Project page `https://yuli0103.github.io/LayoutDiT.github.io/` and its repo: no release assets or checkpoint files.
- Upstream issues #1, #6, #10 (weight release requests; no URL provided).

</details>

An inquiry to the upstream authors requesting the checkpoints is the remaining option for restoring the conversion path; until then the ready-heavy conversion cannot proceed.

**Comment by creative-graphic-design-dev** [comment]
## Plan amendment: train-first LightningCLI and parity staging

CGB-DM remains train-first because no public generator weights are available.
The converted package should use the shared class-path-driven LightningCLI
entry point:

```bash
uv run --package cgb-dm --extra training \
  python -m traingen.lightning.cli fit \
  --config models/cgb-dm/configs/training/cgb_dm_pku_posterlayout.yaml
```

Do not add `cgb_dm.training.cli` in the first PR unless a model-specific
preflight hook is required and documented. YAML files should carry full
`model.class_path` and `data.class_path` values.

The vendor is plain PyTorch, so S0-S2 parity should compare a local vendor-step
adapter against `CGBDMTrainingModule.training_step`. Required traces include
image/pixel values, layout, saliency box, timestep, sampled noise, condition or
fix mask, noised layout, predicted epsilon, loss, and CGB weight. S0 compares
state dicts under a name map plus config/schedule/optimizer defaults.

First PR scope: package skeleton, config/model/scheduler/processor/pipeline CPU
tests, Lightning training skeleton and YAMLs, synthetic/gated S0-S2 parity
hooks, package-local training checkpoint conversion, and from_pretrained smoke.
No full PKU/CGL training, 13 GB dataset download in tests, Hub publishing, or
final parity claim belongs in the first PR.

Public `content_image` is the production equivalent of vendor `uncond`; plain
`unconditional` without image/content should raise. The CGB-DM-specific image,
saliency, saliency-box, and cosine-training-vs-linear-DDIM schedule behavior
stays local. PKU/CGL label ids, including INVALID/pad handling, must be verified
before plan-agreed.

**Comment by creative-graphic-design-dev** [comment]
S0 data-order deviation has been resolved with an explicit regenerated manifest.

Summary:
- The original CGB-DM training loader enumerates `train/inpaint` with `os.listdir`, so row order is filesystem-dependent.
- The package default remains sorted order for deterministic normal data access.
- Vendor parity now captures the observed original-loader filename order into `.cache/cgb-dm/reference/pku_posterlayout_train_manifest.json` and replays that manifest with a vendor-compatible dataset encoding path.
- The manifest is regeneration metadata and is not committed.
- `models/cgb-dm/TRAINING.md` now records this deviation.

Verification on the extracted PKU vendor assets:
- Downloaded and verified the 13.4 GB vendor `Dataset.zip` with `unzip -tq`.
- Extracted PKU under `/dev/shm/cgb-dm/extracted/Dataset/pku/split`; verified train `7734` rows and val `1000` rows.
- Generated the manifest with:
  `uv run --package cgb-dm python models/cgb-dm/scripts/generate_reference_outputs.py --dataset pku_posterlayout --data-root .cache/cgb-dm/datasets/pku/split --manifest-output .cache/cgb-dm/reference/pku_posterlayout_train_manifest.json`
- S0 real-loader parity passed bit-exact for the gated check:
  `PARITY_REQUIRE=1 CGB_DM_DATA_ROOT=.cache/cgb-dm/datasets/pku/split CGB_DM_VENDOR_ORDER_MANIFEST=.cache/cgb-dm/reference/pku_posterlayout_train_manifest.json uv run --package cgb-dm --extra vendor --with pytz pytest models/cgb-dm/tests/vendor_parity -m vendor_parity -q`
  Result: `2 passed`.

S5 full training has not been started from this update. Real S1/S2 training-step trace parity remains the next gate before launching S5.

**Comment by creative-graphic-design-dev** [comment]
## S5 PKU Full-Run Evaluation Comparison

Compared the completed PKU full-training outputs for the CGB-DM S5 evaluation step.

Protocol notes:
- Dataset/protocol: PKU `pku.yaml` training-eval path (`val/inpaint`, 1,000 samples), matching the vendor full-run evaluation path used during training.
- Vendor full-run reference: `pku_full_vendor_20260723_224914`, epoch 500 sampling from the training process log.
- Ours checkpoint: `pku_full_ours_20260724_013039/lightning_logs/version_0/checkpoints/epoch=499-step=121000.ckpt`.
- Additional sampling comparison: reloaded both final checkpoints and generated seeds 1, 2, and 3 with the same evaluation formulas. The standalone vendor seed-1 metrics differ slightly from the training-process log because the training log samples after 500 epochs with the in-process RNG state, while standalone evaluation resets with `set_seed(seed)` after checkpoint load.

| metric | vendor full-run log | vendor standalone mean +/- std (n=3) | ours mean +/- std (n=3) | ours - vendor standalone | ours - vendor log |
|---|---:|---:|---:|---:|---:|
| val | 1.000000 | 1.000000 +/- 0.000000 | 0.089132 +/- 0.003409 | -0.910868 | -0.910868 |
| ove | 0.002727 | 0.002286 +/- 0.000156 | 0.571271 +/- 0.004957 | +0.568984 | +0.568544 |
| undl | 0.996477 | 0.996406 +/- 0.001368 | NaN +/- NaN | NaN | NaN |
| unds | 0.978788 | 0.972385 +/- 0.000736 | NaN +/- NaN | NaN | NaN |
| occ | 0.127215 | 0.127496 +/- 0.000878 | 0.053226 +/- 0.001040 | -0.074271 | -0.073989 |
| rea | 0.015321 | 0.015695 +/- 0.000349 | 0.006598 +/- 0.000607 | -0.009097 | -0.008723 |

Per-seed class distribution check explains the undefined underlay metrics: ours seed 1 generated labels `{0: 14505, 1: 552, 2: 943}` with no class `3` underlay elements, while vendor seed 1 generated `{0: 12267, 1: 2542, 2: 514, 3: 677}`.

Interpretation:
- The full-training ours checkpoint does not match the vendor S5 distribution.
- `unds`/`undl` are undefined for ours because the generated layouts contain no underlay class in the checked seed, so the underlay metric has no support.
- The low `occ` and `rea` values are not an improvement signal here; they are coupled with very low validity and missing underlay/text-heavy structure.

Artifacts kept out of git:
- Ours: `.cache/cgb-dm/full-run/s5-eval-ours-pku-val-fast/summary.json`
- Vendor standalone: `.cache/cgb-dm/full-run/s5-eval-vendor-pku-val-fast/summary.json`
- Per-seed sample tensors and metric JSONs are in the same directories.

Commands run:
```bash
CUDA_VISIBLE_DEVICES=1 uv run --package cgb-dm --extra vendor --with pytz \
  python .cache/cgb-dm/full-run/evaluate_ours_pku_s5.py \
  --backend ours \
  --repo-root "$PWD" \
  --data-root .cache/cgb-dm/datasets/pku/split \
  --checkpoint .cache/cgb-dm/full-run/ours-pku/pku_full_ours_20260724_013039/lightning_logs/version_0/checkpoints/epoch=499-step=121000.ckpt \
  --output-dir .cache/cgb-dm/full-run/s5-eval-ours-pku-val-fast \
  --gpu 0 \
  --seeds 1 2 3

CUDA_VISIBLE_DEVICES=1 uv run --package cgb-dm --extra vendor --with pytz \
  python .cache/cgb-dm/full-run/evaluate_ours_pku_s5.py \
  --backend vendor \
  --repo-root "$PWD" \
  --data-root .cache/cgb-dm/datasets/pku/split \
  --checkpoint .cache/cgb-dm/full-run/vendor-pku/checkpoints/pku_full_vendor_20260723_224914/Epoch500_cgbdm_weights.pth \
  --output-dir .cache/cgb-dm/full-run/s5-eval-vendor-pku-val-fast \
  --gpu 0 \
  --seeds 1 2 3
```

**Comment by creative-graphic-design-dev** [comment]
## S5 architecture-corrected evaluation result

The CGB-DM S5 mismatch has been re-evaluated after the root-cause fixes in `7854247` and the documentation update in `3675b1c`.

Root causes fixed:

- The package `CGBDMTransformerModel` did not match the original `LayoutModel` architecture. The discarded package model had 22.8M parameters and 129 state-dict keys; the original model has 47.9M parameters and 229 keys.
- The earlier S1/S2 parity tests injected the original `LayoutModel` into the package training wrapper, so they did not validate the package model architecture directly.
- The full-run PKU data path used public encoding, which wrote padding as internal class `3` (`underlay`) instead of internal `0`.
- The earlier evaluator used public pipeline decoding before metrics; S5 now evaluates raw internal class ids and boxes with the original metric formulas.

Architecture-corrected S5 evaluation used the same PKU `pku.yaml` validation protocol as before: `val/inpaint`, 1,000 samples per seed, seeds 1/2/3, raw internal `argmax` class ids and boxes.

| Metric | Reference standalone mean +/- std (n=3) | Package mean +/- std (n=3) |
| --- | ---: | ---: |
| `val` | 1.000000 +/- 0.000000 | 1.000000 +/- 0.000000 |
| `ove` | 0.002286 +/- 0.000156 | 0.003293 +/- 0.000801 |
| `undl` | 0.996406 +/- 0.001368 | 0.999345 +/- 0.000284 |
| `unds` | 0.972385 +/- 0.000736 | 0.991428 +/- 0.001467 |
| `occ` | 0.127496 +/- 0.000878 | 0.116661 +/- 0.000648 |
| `rea` | 0.015695 +/- 0.000349 | 0.014180 +/- 0.000295 |

The corrected package checkpoint now matches the reference S5 distribution at the same practical level as standalone reference resampling. The previous low-validity/no-underlay checkpoint is discarded and remains documented only as diagnostic evidence in `models/cgb-dm/TRAINING.md`.

**Comment by creative-graphic-design-dev** [comment]
## CGL S5 rescue and evaluation status

CGL S5 is complete for PR #167 at `26c7d5f1effbe40eab8f21a32199e1f371b1f39c`.

Vendor rescue:
- The previously suspected vendor process IDs `4085147` / `4085415` are no longer running.
- The vendor run had completed successfully: `Epoch500_cgbdm_weights.pth` exists at `/tmp/cgb-dm/full-run/vendor-cgl/cgl_full_vendor_20260725_012722/checkpoints/cgl_full_vendor_20260725_012722/Epoch500_cgbdm_weights.pth`.
- No CGL evaluation or training job is currently running. The rescue S5 jobs completed earlier across GPUs for seeds 1, 2, and 3.

CGL S5 result, package minus reference:

| Metric | Reference mean +/- std (n=3) | Package mean +/- std (n=3) | Delta |
| --- | ---: | ---: | ---: |
| `val` | 0.999097 +/- 0.000109 | 0.999213 +/- 0.000044 | +0.000115 |
| `ove` | 0.001795 +/- 0.000044 | 0.001790 +/- 0.000203 | -0.000005 |
| `undl` | 0.997452 +/- 0.000849 | 0.996399 +/- 0.001292 | -0.001052 |
| `unds` | 0.983453 +/- 0.001646 | 0.987553 +/- 0.002680 | +0.004100 |
| `occ` | 0.115873 +/- 0.000318 | 0.116357 +/- 0.000279 | +0.000484 |
| `rea` | 0.005768 +/- 0.000118 | 0.005971 +/- 0.000115 | +0.000203 |

Verdict: CGL S5 is a practical parity pass. Local artifacts are `.cache/cgb-dm/full-run/s5-eval-ours-cgl-val-summary.json`, `.cache/cgb-dm/full-run/s5-eval-vendor-cgl-val-summary.json`, and `.cache/cgb-dm/full-run/s5-eval-cgl-comparison.json`.

Validation performed before push:
- `uv run --package cgb-dm pytest models/cgb-dm/tests/test_evaluate_full_run.py models/cgb-dm/tests/test_configuration_cgb_dm.py -q` -> 3 passed, 3 warnings.
- `uv run --package cgb-dm ruff check models/cgb-dm/scripts/evaluate_full_run.py models/cgb-dm/tests/test_evaluate_full_run.py` -> passed.
- `git diff --check` -> passed.
- Commit pre-commit hooks -> passed.

Remaining before un-draft:
- PR #167 is still draft.
- GitHub reports `mergeable=CONFLICTING` / `mergeStateStatus=DIRTY` against `main`; no rebase was performed.
- No branch checks are currently reported for `impl-cgb-dm`; rerun/confirm checks after conflict resolution.
